Add translations' test support

jaumef opened the pull request:
Updates:

- [x] Add new translation check method for base test (_destral/testing.py_@OOBaseTest)
- [x] Generate pot file from loaded module using _WizardExportTrans_(scripts/update_trans.py) and...
  - [x] Check all source strings are in module's pot file ( _addons_path/module/i18n/module.pot_ )
  - [x] Check all strings in _es_ES.po_ are in _module.pot_
- [x] Update destral docs
- [x] Use lang from environment:
  - New mandatory env-var: _DESTRAL_TESTING_LANGS_
  - ~~A string with all langs to eval using the character ',' as separator~~
- [x] Only tests if `testing_langs` is not empty

----
Pending checks:

- Report strings must contain comment with mako's name & line
- Report strings must be generated manually
- Report and module strings must be merged before check

ecarreras commented:
@jaumef don't hardcode langs, take it from [detral config environment](https://github.com/gisce/destral/blob/c15f383bed1ea27d46205e6dc05723777ca350bb/destral/testing.py#L20) `DESTRAL_TESTING_LANGS` as a list.

You can update the [documentation](http://destral.readthedocs.io/en/latest/README.html#configuring-destral)

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,104 @@ def test_access_rules(self):
+                pot.write(base64.b64decode(wiz_data))
+
+        pot_path = join(trad_path, '{}.pot'.format(self.config['module']))
+        po_path = join(trad_path, 'es_ES.po')
```
Don't hardcode languages

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,104 @@ def test_access_rules(self):
+        untranslated_strings = compare_pofiles(tmp_pot, po_path, True)
+        assert len(untranslated_ids) == 0
+        assert missing_strings
+        assert untranslated_strings
```
use `self.assertX` instead of `assert`

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,104 @@ def test_access_rules(self):
+                'format': 'po',
+                'modules': [(6, 0, mod_ids)],
+            })
+            wiz_model.act_getfile(cursor, uid, [wiz_id])
```
You can use `tools.trans_export` from openerp code

jaumef commented:
@ecarreras Test is working :smile: :+1:

- Added the environment var "_DESTRAL_TESTING_LANGS_".
- Add a for lang in env-var
- Moved compare_pofiles to utils.py
- Using tools' translate_export to create pot file instead of wizard's
- Using self.asserts instead of asserts
- Removed database's translations check

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,66 @@ def test_access_rules(self):
+            # Generate POT data from loaded strings
+            trans_data = StringIO()
+            trans_export(
+                self.config['testing_langs'][0], self.config['module'],
```
Don't use only the first parameter, you must iterate over **all testing langs**

jaumef replied:
@ecarreras I'd check it for each language, but all POT files are identical, isn't it?

ecarreras replied:
sure, my fault

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,66 @@ def test_access_rules(self):
+                trans_data, 'po', dbname=cursor.dbname
+            )
+
+        tmp_pot = '/tmp/{}.pot'.format(self.config['module'])
```
Use temp files for pot files

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,66 @@ def test_access_rules(self):
+                    missing_strings, self.config['module']
+                )
+            )
+            for test_lang in self.config['testing_langs'].split(','):
```
you don't need to use split, if the environment lang can be parsed as a list it will be a native python list.
e.g.:

```shell
export DESTRAL_TESTING_LANGS="['es_ES', 'en_US']"
```

then you can iterate over the variable

```python
for test_lang in self.config['testing_langs']:
    ...
```

ecarreras commented on destral/testing.py:
```diff
@@ -158,6 +158,66 @@ def test_access_rules(self):
+
+        finally:
+            # Ensure file removal after pofile compare
+            os.system('rm {}'.format(tmp_pot))
```
:bomb: :warning: **Don't use `os.system`!!!**

ecarreras commented on destral/utils.py:
```diff
@@ -144,3 +144,46 @@ def coverage_modules_path(modules_to_test, addons_path):
+    from babel.messages import pofile
+    from os.path import isfile
+    import logging
+    logger = logging.getLogger('Translations')
```
Use root logger

```python
logger = logging.getLogger('destral.utils')
```
or if you want a specific you can do:

```python
logger = logging.getLogger('destral.utils.compare_pofiles')
```

ecarreras commented on destral/utils.py:
```diff
@@ -144,3 +144,46 @@ def coverage_modules_path(modules_to_test, addons_path):
+        logger.info("There are {} strings missing translation in {}".format(
+            not_translated, pathB
+        ))
+    return not_found if not translate else not_translated+not_found
```
For me will be more standard return all the results as a tuple:

eg: tuple:

```python
return not_found, not_translated
```

ecarreras commented on docs/README.rst:
```diff
@@ -50,6 +50,9 @@ Variables definition
   create a temporally database for this test.
+`DESTRAL_TESTING_LANGS`:
+  The langs to be used while checking the translations of each module. This variable
+  is mandatory as it should always be provided.
```
and have to be a list of languages. e.g. `['es_ES', 'en_US']`

ecarreras commented:
A great feature to not break current tests, you can initialize testing_langs to an empty list
```python
self.config = config_from_environment(
    'DESTRAL', ['module'], use_template=True, testing_langs=[]
)
```
Then if `testing_langs` config is empty print a warning but not run the test

ecarreras commented on .gitignore:
```diff
@@ -42,7 +42,6 @@ coverage.xml
 
 # Translations
 *.mo
-*.pot
```
Don't remove *.pot files, I think you can force add to the repo

ecarreras commented:
Closes #73
